exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/Fork.java

```diff
@@ -159,12 +170,38 @@ void rollback() {
     checkState(nativeCanRollback(getNativeHandle()),
         "This fork does not support rollbacks");
 
-    // TODO: Invalidate all indexes created with the fork or the Core won't let us
-    //  do anything.
+    closeDependentObjects();
 
     nativeRollback(getNativeHandle());
   }
 
+  private void closeDependentObjects() {
+    // Clear the registry of opened indexes as they will be closed
+    clearOpenIndexes();
+
+    // Close the active collections (and any other dependent objects),
+    // as rollback requires their invalidation
+    try {
+      forkCleaner.close();
+    } catch (CloseFailuresException e) {
+      // todo: Such situation must not normally happen. Shall we really abort rollback in this case?
+      //   Or shall we proceed with the rollback?
```

Comment:
If we proceed, it probably will lead to RuntimeException from Rust, so failing earlier is better, imo.

Reply:
Not necessarily, the reason I added this comment is that the cleaner is accessible to the client code and it can register there any actions. That makes it possible for the client code to prevent #before_commit invocation for other services. It is not, however, supposed to do that, and in the majority of cases only indexes and other dependent objects will register their destructors in the cleaner.

Reply:
That's interesting. What about throwing RuntimeException then? If the indexes weren't closed correctly, it will arise anyway, but if the user modifies cleaners - exception will be the sign of disapprove.

Reply:
The exception type must not make a difference, as any exceptions from destructors are caught by the cleaner and are presented as a CloseFailuresException. Therefore, if we consider this situation as (a) unlikely and (b) not smth we *must* prevent (as there are already enough ways to break things), we may leave it as is (consider a failure coming from our destructors which *must not* happen).